### PR TITLE
fix(css): Improve formatting of at-rules and add tests

### LINKED ISSUE #3875: Build is failing on Node.js 4 in production mode
Several recent testing runs have failed because the Jest process gets “Killed:”

```
Killed
error Command failed with exit code 137.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
/home/travis/build/prettier/prettier/node_modules/shelljs/src/common.js:417
      if (config.fatal) throw e;
                        ^
Error: exec: internal error
    at Object.error (/home/travis/build/prettier/prettier/node_modules/shelljs/src/common.js:128:27)
    at Object._exec (/home/travis/build/prettier/prettier/node_modules/shelljs/src/exec.js:292:12)
    at Object.exec (/home/travis/build/prettier/prettier/node_modules/shelljs/src/common.js:350:23)
    at Object.<anonymous> (/home/travis/build/prettier/prettier/scripts/test-dist.js:23:20)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
    at startup (node.js:140:18)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

[source](https://travis-ci.org/prettier/prettier/jobs/335759276#L1087-L1106)

### PR DESCRIPTION
**Pull request by evilebottnawi**
Fixes #3875 #3721 #3282 #2852 #2594 #2158

Feel free to review :+1:

### CODE REVIEW

File: src/language-css/parser-postcss.js [hunk 1]
```diff
@@ -208,26 +214,78 @@ function parseNestedCSS(node) {
...
-    if (node.type === "css-atrule" && typeof node.params === "string") {
+
+    if (
+      node.type === "css-atrule" &&
+      typeof node.params === "string" &&
+      node.params.trim().length > 0
+    ) {
+      if (
+        node.name.toLowerCase() === "counter-style" ||
+        node.name.toLowerCase().endsWith("keyframes") ||
```
Comment (lydell) [inline src/language-css/parser-postcss.js]: Why does this one need `.endsWith()`?

File: src/language-css/parser-postcss.js [hunk 2]
```diff
@@ -208,26 +214,78 @@ function parseNestedCSS(node) {
...
-          return {
-            type: "media-unknown",
-            value: node.params
-          };
+
+        return node;
+      }
+
+      if (node.name === "extend") {
+        node.selector = parseSelector(node.params);
```
Comment (lydell) [inline src/language-css/parser-postcss.js]: This is so good 👍

File: src/language-css/parser-postcss.js [hunk 3]
```diff
@@ -208,26 +214,78 @@ function parseNestedCSS(node) {
...
+        delete node.params;
+
+        return node;
+      }
+
+      if (node.name === "at-root") {
+        if (/^(\(\s*(without|with)\s*:\s*.+\s*\)$)/.test(node.params)) {
```
Comment (lydell) [inline src/language-css/parser-postcss.js]: `/^\(\s*(without|with)\s*:[\s\S]+\)$/`?

Can there be whitespace at the start and end? Or is that something the parser has taken care of?

File: tests/css_scss/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -856,7 +856,11 @@ p {
 @function myCalculationFunction($some-number, $anotherNumber, $BigNumber) {
   @return $some-number + $anotherNumber + $BigNumber;
 }
-@function AnotherMyCalculationFunction($some-number, $anotherNumber, $BigNumber: 100px) {
+@function AnotherMyCalculationFunction(
+    $some-number,
+    $anotherNumber,
+    $BigNumber: 100px
+  ) {
```
Comment (lydell) [inline tests/css_scss/__snapshots__/jsfmt.spec.js.snap]: Should this be:

```scss
@function AnotherMyCalculationFunction(
  $some-number,
  $anotherNumber,
  $BigNumber: 100px
) {
  @return $some-number + $anotherNumber + $BigNumber;
}
```

?

File: src/language-css/parser-postcss.js (hunk 1 above)
Comment (j-f1) [reply src/language-css/parser-postcss.js]: It’s probably for prefixed definitions, like `-webkit-keyframes`.

File: tests/css_scss/__snapshots__/jsfmt.spec.js.snap (hunk 4 above)
Comment (j-f1) [reply tests/css_scss/__snapshots__/jsfmt.spec.js.snap]: (i.e. remove a space from the beginning of the parameter lines)

File: src/language-css/parser-postcss.js (hunk 1 above)
Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: @lydell prefixes `-webkit-keyframes` and etc

File: src/language-css/parser-postcss.js (hunk 3 above)
Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: @lydell valid with any whitespace (`\t`, `\n`, `  ` and etc)

File: tests/css_scss/__snapshots__/jsfmt.spec.js.snap (hunk 4 above)
Comment (evilebottnawi) [reply tests/css_scss/__snapshots__/jsfmt.spec.js.snap]: @lydell yep, first all at rules and tests, second formatting with feedback

Comment (evilebottnawi) [reply tests/css_scss/__snapshots__/jsfmt.spec.js.snap]: it is still WIP :smile: thanks for feedback, i ping you when all looks good and tested

File: src/language-css/parser-postcss.js (hunk 1 above)
Comment (lydell) [reply src/language-css/parser-postcss.js]: Thanks! :+1:

File: src/language-css/parser-postcss.js (hunk 3 above)
Comment (lydell) [reply src/language-css/parser-postcss.js]: I don't understand what you mean?

Comment (evilebottnawi) [reply src/language-css/parser-postcss.js]: @lydell valid:
```scss
@at-root 

(

without

: 

media

) {
    color: red;
}
```

Comment (lydell) [reply src/language-css/parser-postcss.js]: Let me rephrase:

1. What do you think of my Regex suggestion? `/^\(\s*(without|with)\s*:[\s\S]+\)$/`
2. Should there `\s*` at the start and end? `/^\s*\(\s*(without|with)\s*:[\s\S]+\)\s*$/`

**Comment by evilebottnawi** [comment]
@lydell @j-f1 ready for feedback around improve stylistics at rules, all `css/scss/less` at rule tests added and passed

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -1,98 +1,2394 @@
...
+
+    @at-root (with: media) {
+      color: red;
+    }
+  }
+}
+@media print {
+  .page {
+    width: 8in;
+    @at-root (with: media) {
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Would it make sense to always enforce a bank line before `@at-root`?

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -1,98 +1,2394 @@
...
+h6
+
+;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+@custom-selector :--heading h1,
+  h2,
+  h3,
+  h4,
+  h5,
+  h6;
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: How about:

```scss
@custom-selector :--heading (
  h1,
  h2,
  h3,
  h4,
  h5,
  h6
);
```

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 7]
```diff
@@ -1,98 +1,2394 @@
...
+
++
+
+12em
+
+;
+@debug $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var + $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+@debug 10em + 12em;
+@debug 10em+12em;
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]:
```scss
@debug 10em + 12em;
```

Or would that be too difficult to support?

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 8]
```diff
@@ -1,98 +1,2394 @@
...
+@debug 10em + 12em;
+@debug 10em + 12em;
+@debug 10em + 12em;
+@debug 10em + 12em;
+@debug $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var +
+  $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var;
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]:
```scss
@debug (
  $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var +
  $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var
);
```

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 9]
```diff
@@ -1,98 +1,2394 @@
...
+}
+@each $animal, $color, $cursor in (puma, black, default),
+   (sea-slug, blue, pointer),  (egret, white, move) {
+}
+@each $animal, $color, $cursor in (puma, black, default),
+  (sea-slug, blue, pointer), (egret, white, move) {
+}
+@each $animal, $color, $cursor in (puma, black, default),
+   (sea-slug, blue, pointer),  (egret, white, move) {
+}
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]:
```scss
@each $animal, $color, $cursor in (
  (puma, black, default),
  (sea-slug, blue, pointer),
  (egret, white, move)
) {
}
```

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 10]
```diff
@@ -1,98 +1,2394 @@
...
+a.important {
+  @extend .notice !optional;
+}
+a.important {
+  @extend .notice !optional;
+}
+.message-error {
+  @extend .message,
+    .important;
```
Comment (j-f1) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]:
```scss
@extend .message, .important;
```

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 5 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 i think add blank line is bad practice, i am working in many project where people don't add newlines before at rules and always add newline before at rules. I think better output as is. If many contributors vote for the newline can add in the future.

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 6 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 invalid syntax, spec don't allow add `(` and `)`

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 7 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 no, problems in https://github.com/prettier/prettier/blob/master/src/language-css/printer-postcss.js#L341, when i use two space or newlines all look good, seems it is separate issue, i do PR and when rebase

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 10 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 yep, also will add test when selector is very long

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 9 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 do you think it normal to add `(` and `)` if they are not necessary (we don't output `(` and `)` they are unnecessary in js)?

**Comment by evilebottnawi** [comment]
@j-f1 about extension - i think no, we use `IS_POSSIBLY_SCSS` and it is allow tests what syntax correctly parsed

**Comment by lydell** [comment]
@evilebottnawi `IS_POSSIBLY_SCSS` is kinda broken. If we can avoid it it’d be great.

**Comment by evilebottnawi** [comment]
@lydell hm let's open separate Issue/PR for this, i don't think it's worth putting it all in one PR

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 10 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @j-f1 done

**Comment by j-f1** [comment]
@evilebottnawi FYI you don’t have to squash your commits before you merge — GitHub will do that for you when you click “Squash and merge.”

**Comment by evilebottnawi** [comment]
@j-f1 i known, habit, not all repos on github/gitlab

**Comment by evilebottnawi** [comment]
@lydell @j-f1 friendly ping, some questions:
1. Many of `scss` structure don't support parens around control directive (example `@for`), i.e. i think better avoid using parens where is possible. Also scss guidelines advice don't use parens too. But remove unnecessary parens i will do in next PR. This PR already very big and difficult.
2. Should we add breakline in multi line control directive.
Input:
```scss
@while $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var > $other-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var
{
}
```
Output:
```scss
@while 
  $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var >
  $other-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var
{
}
```
We always move on new line variables in control directive into `js`
3. Some tests contain `$i>0`, it is also net related to this PR and will do in other next PR (problem in parse `$i>0` - is `value-word`)
4. Difficult `@supports` directives look very uglify, problem in `postcss-media-parser`. It should be fixed in parser. In near future i do PR
5. `$args ...` will be do in other PR, it is don't break code
6. `mix (...)` parses as `value-word` + `parens` + '...' + `parent`. Css don't support here space, but scss support. Will be fixed in other PR. It is also don't break code.
7. `@for` directive
Input:
```scss
@for $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var from ($very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var1 + $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var1) through ($very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var-2 + $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var-2) {}
```
Output:
```scss
@for $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var
  from
  (
    $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var1 +
      $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var1
  )
  through
  // Should we add indent after keywords when they multi line?
  (
    $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var-2 +
      $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var-2
  )
{
}
```
8. Also please look on `@custom-selector` and wait feedback.
Thanks for everyone!

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 11]
```diff
@@ -1,98 +1,2553 @@
...
+}
+@each $animal in ( (puma),  (sea-slug),  (egret),  (salamander)) {
+}
+@each $animal in ((puma),  (sea-slug),  (egret),  (salamander)) {
+}
+@each $animal in ( (puma),  (sea-slug),  (egret),  (salamander)) {
+}
+@each $animal in ( (puma),  (sea-slug),  (egret),  (salamander)) {
+}
+@each $animal in ( (puma),  (sea-slug),  (egret),  (salamander)) {
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Some extra spaces seem to be printed here. Just in case you aren't already aware of these.

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 12]
```diff
@@ -1,98 +1,2553 @@
...
+{
+}
+@each $animal, $color, $cursor in (puma, black, default),
+   (sea-slug, blue, pointer),  (egret, white, move)
+{
+}
+@each $very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-var
+    in
+    puma,
+  sea-slug, egret, salamander
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Isn't the indentation odd here?

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 13]
```diff
@@ -1,98 +1,2553 @@
...
+  }
+  @if $type == ocean {
+  }
+  @if $type == ocean {
+  }
+  @if $type == ocean {
+  }
+  @if (($type) == (ocean)) {
+  }
+  @if (($type)==(ocean)) {
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: This `==` is not spaced. (Same thing for some other cases a couple of lines down.)

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 14]
```diff
@@ -1,98 +1,2553 @@
...
+    color: green;
+  } @else {
+    color: black;
+  }
+  @if $very-very-very-long-var == 0 and $very-very-very-long-var == 0 {
+  }
+  @if $very-very-very-very-very-very-long-var ==
+    0 and
+    $very-very-very-long-var ==
+    0
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Ideally, the `0`s should be on the same line as the `==`s.

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 15]
```diff
@@ -1,98 +1,2553 @@
...
+  @if $very-very-very-very-very-very-very-very-very-very-very-long-var ==
+    0 and
+    $very-very-very-very-very-very-very-very-very-very-very-long-var ==
+    0
+  {
+  }
+  @if $base-font-size !=
+    16px or
+    $base-line-height !=
+    24px or
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Same thing here: Would be great if we could keep the comparisons together.

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap [hunk 16]
```diff
@@ -308,6 +3193,12 @@ run-in
 }
 @supports not (transform-origin: 10em
 10em
+10em) {
+}
```
Comment (lydell) [inline tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: Something has happened here with all the newlines and stuff.

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 14 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell yep need fix :+1:

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 15 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell :+1:

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 16 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell problem in parser, i will fix this in parser, but it is valid code

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 13 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell :+1:

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 12 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell yep, fix here :+1:

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 11 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell yep, known, will be fixed in other PR with fix spaces around `(` and `)`

File: tests/css_atrule/__snapshots__/jsfmt.spec.js.snap (hunk 13 above)
Comment (evilebottnawi) [reply tests/css_atrule/__snapshots__/jsfmt.spec.js.snap]: @lydell problem in `postcss-values-parser`. Issue https://github.com/shellscape/postcss-values-parser/issues/42.

**Comment by evilebottnawi** [comment]
Thanks to everyone! In near future i can do PR in parsers and fix some uglify places.

**Comment by j-f1** [comment]
FYI the reason the issues weren’t automatically closed is because you have to put “fixes” before each issue reference, i.e.

```markdown
fixes #1 fixes #2 fixes #3
```

**Comment by evilebottnawi** [comment]
@j-f1 yep, accidentally incorrectly written `fixes`

**Comment by j-f1** [comment]
Don’t worry, it’s not that obvious unless you read [the docs 📖](https://help.github.com/articles/closing-issues-using-keywords/#closing-multiple-issues) :)
